Fixes #25487 - Fix subtotal in API response

johnpmitsch opened the pull request:
When requesting an endpoint with a search parameter, the subtotal
shows the results returned, rather than responses found. i.e. even
if 200 results were found, if 20 are being returned because of
pagination, the subtotal will be 20.

This changes back to using `count`, which sends another SQL query
and returns the correct subtotal.

It looks like this was updated (with good intention) for performance reasons https://github.com/theforeman/foreman/pull/5993/files#r213566553


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

Commit:
Fixes #25487 - Fix subtotal in API response

When requesting an endpoint with a search parameter, the subtotal
shows the results returned, rather than responses found. i.e. even
if 200 results were found, if 20 are being returned because of
pagination, the subtotal will be 20.

This changes back to using `count`, which sends another SQL query
and returns the correct subtotal.

johnpmitsch commented on app/controllers/api/v2/base_controller.rb:
```diff
@@ -61,12 +61,12 @@ def root_node_name
 
       def metadata_total
-        @total ||= resource_scope.try(:size).to_i
+        @total ||= resource_scope.try(:count).to_i
```
This total is correct (afaict), should we use count here as well?

tbrisker replied:
I think here `size` should be safe and save one query possibly, but that's what tests are for

johnpmitsch commented:
credit to @iNecas and @mbacovsky for the investigation :mag:

iNecas commented:
It would be good to add some test to cover this behavior (I'm sure we would not hit this issue, if there was some testing for this before).

Commit:
Refs #25487 - added tests

johnpmitsch commented:
@iNecas I added testing. I figured the best way to test the metadata is a real endpoint, so I added to the host controller tests. Let me know if there is a better way

tbrisker commented on test/controllers/api/v2/hosts_controller_test.rb:
```diff
@@ -1230,4 +1230,33 @@ def setup
+  describe 'should create correct subtotals' do
+    before do
+      as_admin do
+        100.times { |_| FactoryBot.create(:host) }
```
you can use `FactoryBot.create_list` instead

tbrisker commented on test/controllers/api/v2/hosts_controller_test.rb:
```diff
@@ -1230,4 +1230,33 @@ def setup
+      hosts_response = ActiveSupport::JSON.decode(@response.body)
+
+      assert_response :success
+      assert_equal hosts_response["subtotal"], ::Host.all.count
```
I don't think the  `::` is needed here

tbrisker reviewed (changes requested):
good catch! me-- for introducing this. A couple of minor comments inline.

Commit:
Refs #25487 - Updated tests to include total
and use size for total

johnpmitsch commented:
@tbrisker thanks for the suggestions, I updated the tests to test for total too. I set up the hosts under "Empty org" so the subtotal and total are different for the test with the search parameter. I changed the metadata_total to use `size` again and it is still passing the tests :+1:

johnpmitsch commented:
[test foreman]

johnpmitsch commented:
@tbrisker does this approach look ok to you?

iNecas reviewed (dismissed):
Works for me

tbrisker commented on test/controllers/api/v2/hosts_controller_test.rb:
```diff
@@ -1230,4 +1230,36 @@ def setup
+
+      assert_response :success
+      assert_equal hosts_response["subtotal"], @empty_org.hosts.count
+      assert_equal hosts_response["total"], Host.all.count
```
There is a hidden assumption here, that not all hosts are in `@empty_org` - perhaps assert that the subtotal and total are different?

tbrisker commented on test/controllers/api/v2/hosts_controller_test.rb:
```diff
@@ -1230,4 +1230,36 @@ def setup
+    before do
+      @empty_org = Organization.find_by_name("Empty Organization")
+      as_admin do
+        FactoryBot.create_list(:host, 100, organization: @empty_org)
```
instead of creating 100 entries, you can create a smaller list (say 10) and pass a smaller per_page parameter in the request (say 5)

tbrisker reviewed (commented):
sorry for dropping the ball here, just a couple of small inline comments

johnpmitsch commented:
@tbrisker thanks, updated according to your suggestions

tbrisker reviewed (approved):
Thanks @johnpmitsch , lgtm!

iNecas commented:
Tests failures unrelated, Thanks @johnpmitsch and @tbrisker

iNecas commented:
Cherry-pick PR here https://github.com/theforeman/foreman/pull/6299
